keps/sig-windows/4802-windows-node-shutdown/README.md

```diff
@@ -0,0 +1,889 @@
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+*   As a cluster administrator, I can configure the nodes in my cluster to
+    allocate X seconds for my pods to terminate gracefully during a node
+    shutdown
+
+#### Story 2
+
+*   As a developer I can expect that my pods will terminate gracefully during
+    node shutdowns
+
+### Background on Windows Shutdown
+
+In the context of this KEP, shutdown is referred to as shutdown or restart of the
+underlying Windows machine. A shutdown can be initiated via a
+variety of methods for example:
+
+1. `shutdown /t 0`
+2. `shutdown /t 30` `#schedule a delayed shutdown in 30 seconds`
+3. `shutdown /r /t 30` `#schedule a delayed restart in 30 seconds`
+4. Physically pressing the power button on the machine
+5. If a machine is a VM, the underlying hypervisor can press the “virtual”
+   power button
+6. For a cloud instance, stopping the instance via Cloud API, e.g. via `az vm stop -g xx -n xx`.
+   Depending on the cloud provider, this may result in virtual power button press by the underlying hypervisor
+7. Windows update (ex: Patch Tuesday) can also trigger a Windows node to be shutdown and rebooted
+
+In the context of kubernetes on Windows node, system is not
+aware of the pods and containers running on the machine and will simply
+kill them as regular Windows processes, that's the reason that we will need
+to monitor the system for shutdown events and behave accordingly.
+
+The high-level process of detecting a system shutdown on Windows is the same for all types of Windows applications and services. To detect a shutdown, 
+we create a callback function and register it in the system. When a certain Windows event occurs, the system calls the callback function, transferring 
+information about the event via input parameters. The callback function then analyzes the input parameter data, and executes code accordingly. In the 
+case of this feature related, the call back can detect the related (pre)shutdown event, and kill the pods gracefully before exiting
+
+
+There’s no one way to make all types of applications (console, GUI and Windows services) detect a node shutdown, different types of applications differ 
+in their syntax and functionality.
+
+For general Windows console application, it can process the "CTRL_SHUTDOWN_EVENT", but according to [HandlerRoutine callback function](https://learn.microsoft.com/en-us/windows/console/handlerroutine#parameters)
+and the [SetConsoleCtrlHandler function](https://learn.microsoft.com/windows/console/setconsolectrlhandler?redirectedfrom=MSDN) 
+`Console functions, or any C run-time functions that call console functions, may not work reliably during processing of any of the three signals mentioned previously. 
+The reason is that some or all of the internal console cleanup routines may have been called before executing the process signal handler`.
```

Comment:
Is this the reason why the feature cannot be supported when kubelet is running from the CLI? I suggest explicitly calling that out here.

Reply:
I am fine with that. How about we call that out in the beta graduation section?

Reply:
Will keep it here and update later during beta graduation

Reply:
There is a bit of confusion here. If we are going to tackle allowing this feature when running in CLI as pre-req for beta graduation then we should not mention the issue here nor in non-goals. Or are we saying that we will decide if this is even possible only in beta and is not a criteria for graduation?

Reply:
I was suggesting that we will decide if this is even possible before graduating to beta / not worrying about non-cli use cases for the (initial) alpha implementation.

Reply:
Sounds good.